docs/docs/reference/symmetric-meta-programming.md

```diff
@@ -0,0 +1,747 @@
+
+Quotes and splices are primitive forms in the generated abstract
+syntax trees. They are eliminated in an expansion phase
+`ReifyQuotes`. This phase runs after typing and pickling.
+
+Macro-expansion works outside-in. If the outermost scope is a splice,
+the spliced AST will be evaluated in an interpreter. A call to a
+previously compiled method can be implemented as a reflective call to
+that method. With the restrictions on splices that are currently in
+place that’s all that’s needed. We might allow more interpretation in
+splices in the future, which would allow us to loosen the
+restriction.  Quotes in spliced, interpreted code are kept as they
+are, after splices nested in the quotes are expanded.
+
+If the outermost scope is a quote, we need to generate code that
+constructs the quoted tree at run-time. We implement this by
+serializing the tree as a Tasty structure, which is stored
+in a string literal. At runtime, an unpickler method is called to
+deserialize the string into a tree.
+
+Splices inside quoted code insert the spliced tree as is, after
+expanding any quotes in the spliced code recursively.
+
+## Formalization
+
+The phase consistency principle can be formalized in a calculus that
+extends simply-typed lambda calculus with quotes and splices.
+
+### Syntax
+
+The syntax of terms, values, and types is given as follows:
+
+    Terms         t  ::=  x                 variable
+                          (x: T) => t       lambda
+                          t t               application
+                          ’t                quote
+                          ~t                splice
+
+    Values        v  ::=  (x: T) => t       lambda
+                          ’q                pure quote
+
+    Quoted        q  ::=  x  |  (x: T) => q  |  q q  |  ’t
+
+    Types         T  ::=  A                 base type
+                          T -> T            function type
+                          expr T            quoted
+
+Typing rules are formulated using a stack of environments
+`Es`. Individual environments `E` consist as usual of variable
+bindings `x: T`. Environments can be combined using the two
+combinators `’` and `~`.
+
+    Environment   E  ::=  ()                empty
+                          E, x: T
+
+    Env. stack    Es ::=  ()                empty
+                          E                 simple
+                          Es * Es           combined
+
+    Separator     *  ::=  ’
+                          ~
+
+The two environment combinators are both associative with left and
+right identity `()`. The initial environment contains among other
+predefined operations also a "run" operation `! : expr T => T`. That
+is, `!` is treated just like any other function.
+
+### Operational semantics:
+
+We define a small step reduction relation `-->` with the following rules:
+
+                ((x: T) => t) v  -->  [x := v]t
+
+                          ~(’t)  -->  t
+
+                              t  -->  t’
```

Comment:
We should avoid using both `'` for quote and `'` prime. Maybe we should use unstead if we use `t --> t2`.